PetscDMWrapper Part 1

pbauman opened the pull request:
This implements Step 1 in the discussion in #1567.

I've marked this as do not merge until I can add testing into both libMesh and GRINS. I did testing way back when I wrote this code, but it was never formalized.

The testing to be added here is 1. The command line options for fieldsplit in this code path (both serial and parallel) and 2. recursive fieldsplit that I believe the current interface doesn't support. I'm thinking the Laplace/LAGRANGE_VEC example (where we have 3 decouple Laplace equations) and a incompressible thermal flow regression in GRINS. Also, I need to check the min PETSc version. Changelogs in PETSc aren't super helpful. I recall starting this work at PETSc 3.6.X, so was thinking I'll try that. Any urgent need to test earlier? Otherwise I'll just restrict this to PETSc 3.6.4 or later.

And dammit I just noticed I forgot documentation on the `init_and_attach_petscdm` function so I need to add that too.

Anyway, want to let folks start commenting on the design while I get tests into build system since I'm pretty confident this code will work correctly (famous last words...).

pbauman commented:
And also need to setup the stub correctly in the first commit (the .C stub is missing and sucked into 
e5e83e3 which will cause 5240870 to not compile).

jwpeterson commented:
Is there any relation between this new DM stuff and the existing DM stuff ([1](https://github.com/libMesh/libmesh/blob/master/src/solvers/petscdmlibmesh.C), [2](https://github.com/libMesh/libmesh/blob/master/src/solvers/petscdmlibmeshimpl.C))? For example, could we get rid of the old stuff?

pbauman commented:
I don't interact at all with the old stuff. I don't use it (I'll use the stuff in this PR), but does MOOSE use it? I would suggest it can be removed if not, but we should probably poll libmesh-users before we remove it since maybe the original authors might still be using it in some context.

jwpeterson commented:
@fdkong may be interested in this.

pbauman commented:
OK, rebased on master, and fixed the stub commit so all commits in the PR will compile. Still need testing and documentation for `init_and_attach_petscdm`.

fdkong commented on src/solvers/petsc_dm_wrapper.C:
```diff
@@ -0,0 +1,388 @@
+      PetscSection & section = this->get_section(level);
+      PetscSF & star_forest = this->get_star_forest(level);
+
+      ierr = DMShellCreate(system.comm().get(), &dm);
```
I was wondering why we do not implement a native libmesh DM? The native DM will give us more flexibility on the setup of the solver and the preconditioner.  In PETSc, we have native DMs for DMPlex and DMDA. This allows us do more customized stuffs in the future. 

We could clean up the old DM stuff, and combine with this PETScDMWrapper. It should be easy.

I miss something here?

pbauman replied:
> I was wondering why we do not implement a native libmesh DM instead of shell dm?

Can you elaborate on what you mean by "native" DM? Do you mean to essentially do a full subclassing of DM?

> The native DM will give us more flexibility on the setup of the solver and the preconditioner. In PETSc, we have native DMs for DMPlex and DMDA. This allows us do more customized stuffs in the future.

Can you elaborate on what flexibility and customization is gained vs. the shell? As far as I can tell, we will get the full solver flexibility (once the part 2 comes with GMG). What else would be enabled that not's enabled by the shell?

pbauman replied:
The other reason I like the shell interface is it minimizes the PETSc API we have to touch since PETSc has no intent so support backward compatibility in API so the more API we touch, the larger maintenance headache it becomes.

fdkong replied:
> Can you elaborate on what you mean by "native" DM? Do you mean to essentially do a full subclassing of DM?

I mean something like  https://github.com/petsc/petsc/blob/65b40d628b4e55462e59718b6e7e25b2e90f85aa/include/petsc/private/dmpleximpl.h#L175

We may have something similar: DM_Libmesh.

> 
> The native DM will give us more flexibility on the setup of the solver and the preconditioner. In PETSc, we have native DMs for DMPlex and DMDA. This allows us do more customized stuffs in the future.
> Can you elaborate on what flexibility and customization is gained vs. the shell? As far as I can tell, we will get the full solver flexibility (once the part 2 comes with GMG). What else would be enabled that not's enabled by the shell?

Personally I like a native DM more than the shell DM because I could add some internal data (or other crazy things definitely at the user-side) in "typedef struct" that may be useful for implementing the solver interfaces. For example, to construct a geometric DDM, we may need to implement the interface: dm->ops->createdomaindecomposition. If we have our own internal data (such mesh info or system) inside of DM, it is straightforward to feedback the indices to PETSc, otherwise, we have to figure out a way (maybe PetscObjectCompose()) to bring in mesh or system info because the DDM interface does not have AppCtx.

Again, I could be wrong. All these stuffs are based on my experiences. I implemented GMG (a pure petsc-based code) for my  PhD thesis using the native DM.

pbauman replied:
OK, so it looks like the existing DM bits were moving towards a "native" DM.

> Personally I like a native DM more than the shell DM because I could add some internal data (or other crazy things definitely at the user-side) in "typedef struct" that may be useful for implementing the solver interfaces. For example, to construct a geometric DDM, we may need to implement the interface: dm->ops->createdomaindecomposition. If we have our own internal data (such mesh info or system) inside of DM, it is straightforward to feedback the indices to PETSc, otherwise, we have to figure out a way (maybe PetscObjectCompose()) to bring in mesh or system info because the DDM interface does not have AppCtx.

Shell provides access to `createdomaindecomposition` and takes a context (step 2 of this process uses that context hook) so we can feed `EquationSystem` and therefore everything that would be needed. http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/DM/DMShellSetContext.html

That said, I'm not opposed to a native DM, but what's there looks rather cumbersome. Perhaps we could talk off thread about the DM API and what you're looking for. But in the near term, I'd prefer to stick to a shell so we can keep this moving.

fdkong commented on src/solvers/petsc_diff_solver.C:
```diff
@@ -350,18 +352,29 @@ void PetscDiffSolver::setup_petsc_data()
-  ierr = KSPGetPC(my_ksp, &my_pc);
-  LIBMESH_CHKERR(ierr);
+      PC my_pc;
+      ierr = KSPGetPC(my_ksp, &my_pc);
```
It is not introduced by this PR.But  it would be great if we could get ride of this manual setup of the preconditoner.

pbauman replied:
My intent is to eventually (after several libMesh releases) remove the old `petsc_auto_fieldspilt` in favor of this.

fdkong replied:
In MOOSE, we have our own fieldsplit preconditioner implementation. The preconditioner allows us split the matrix based on not only the physics field but also the geometry. This requires us to implement dm->ops->createfielddecomposition.   

I am guessing you are using the default petsc implementation for this. Then we can not construct  the fieldsplit preconditioner based on the geometry.

pbauman replied:
You are correct, the field split access here is only for physics field split. Again, perhaps we (and whoever is interested) can talk off thread about how you use the domain decomposition so we can integrate that cleanly because, I believe, this would provide all you'd need to for a physics field split and not require anything downstream in MOOSE.

fdkong commented:
@pbauman Could you possibly add an example illustrating the usage of FieldSplit based on the PETScWrapper in this PR?  I guess GMG is not working yet in this PR.

pbauman commented:
> Could you possibly add an example illustrating the usage of FieldSplit based on the PETScWrapper in this PR?

Yes, forthcoming (hopefully tomorrow) as mentioned in the PR.

> I guess GMG is not working yet in this PR.

Correct, this is only step 1. Please see #1567 for a more elaborate/global discussion of where I'm going.

pbauman commented:
> Yes, forthcoming (hopefully tomorrow) as mentioned in the PR.

I do have a working test case comparing to PETSc ex62, just needs to be added to GRINS and ought to be able to adapt to one of the fem_system examples.

pbauman commented:
@roystgnr pointed out `dof_owner` should be able to replace my `find_dof_rank` function.

pbauman commented on src/solvers/petsc_dm_wrapper.C:
```diff
@@ -19,4 +19,68 @@
+{
+  const MeshBase & mesh = system.get_mesh();
+
+  unsigned int pstart = 2^30;
```
We need to make this not restricted

pbauman commented on src/solvers/petsc_dm_wrapper.C:
```diff
@@ -19,4 +19,68 @@
+      pend = 0;
+    }
+
+  PetscErrorCode ierr = PetscSectionSetChart(section, pstart, pend);
```
I need to ask @knepley if the chart really want there to be number of points = pend-pstart since we could easily have pstart = 0, pend=1e8 and have only those 2 elements on the processor.

fdkong replied:
We usually assume that the number is consecutively increased from pstart to pend.

pbauman replied:
OK, just finished talking to @knepley. He said it does need to be contiguous, but it doesn't need to be global. In fact, he said because the PetscSF stashes the tuple of local id, local proc id, they don't care really about the point number for solvers (which is all we're using this for). So we can manually construct our own local point numbering and it's fine as long as we're consistent when we set the dof ids for the point. I'll take a shot at working that in here before we merge.

roystgnr replied:
Oh, seriously?  That sounds just fine, then.  Thank him for me.

We're probably going to be stuck creating a lookup table, aren't we?  libMesh ids are all global, but (except for dof numbers) aren't necessarily contiguous.

roystgnr commented:
:+1: after the dof_owner and unsigned-int-vs-PetscInt stuff is fixed, but before committing put up issues for fixing all the current half-baked bits: the "No DoFs on elements" assumption, the "No more than 1 DoF on each node" assumption.  None of those things should be hard to fix, but the question of what is the best way to fix them will depend on the PetscSectionSetChart issue.

roystgnr commented:
Oh, and wait a second - what do we need to do about SCALAR variables here?

pbauman commented:
> Oh, and wait a second - what do we need to do about SCALAR variables here?

Oh, good point. Those ought to be able to be their own "points" and they get 1 for each one. They always live on the processor with the largest rank IIRC? But you're correct, we'll need to account for those as well.

pbauman commented:
I need to not to forget to put PETSc version guards around this too.

bboutkov commented:
As per our discussion, we also want init_and_attach_petscdm to pass in a non-const System as we will be walking the mesh hierarchy.

pbauman commented:
I'm closing this for now because I'm going to working on this and push repeatedly with branch cleanups etc and don't want unnecessarily keep triggering CI. I'll reopen once I'm done.
